qiskit/transpiler/passes/optimization/template_matching/template_substitution.py

```diff
@@ -62,6 +64,16 @@ def __init__(self, max_matches, circuit_dag_dep, template_dag_dep):
         self.unmatched_list = []
         self.dag_dep_optimized = DAGDependency()
         self.dag_optimized = DAGCircuit()
+        self.cost_dict = {'id': 0, 'x': 1, 'y': 1, 'z': 1, 'h': 1, 't': 1, 'tdg': 1, 's': 1,
```

Comment:
These costs seem to be based on default decompositions defined for the gate. Occasionally the default decompositions change or a different target basis is used. I wonder if there is a way to make this initialization more aware of this?

Reply:
@ewinston I think there might be a way to do this using the `SessionEquivalenceLibrary` in Qiskit. One could e.g. do:

```
SessionEquivalenceLibrary = EquivalenceLibrary(base=StandardEquivalenceLibrary)
```
then this
```
iswap = SessionEquivalenceLibrary.get_entry(iSwapGate())[0]
```
gives you the decomposition of iswap into CNOT gates (or whatever mapping you have defined in the equivalence library). Then we could use these circuits to arrive at the `cost_dict` by counting the number of gates. Thoughts?